base_layer/core/src/base_node/sync/block_sync/synchronizer.rs

```diff
@@ -0,0 +1,243 @@
+    }
+
+    async fn get_next_sync_peer(&mut self) -> Result<PeerConnection, BlockSyncError> {
+        match self.sync_peer {
+            Some(ref peer) => Ok(peer.clone()),
+            None => {
+                let mut peers = self
+                    .connectivity
+                    .select_connections(ConnectivitySelection::random_nodes(1, vec![]))
+                    .await?;
+                if peers.is_empty() {
+                    return Err(BlockSyncError::NoSyncPeers);
+                }
+                Ok(peers.remove(0))
+            },
+        }
+    }
+
+    async fn attempt_block_sync(&mut self, mut conn: PeerConnection) -> Result<(), BlockSyncError> {
+        let mut client = conn
+            .connect_rpc_using_builder(rpc::BaseNodeSyncRpcClient::builder().with_deadline(Duration::from_secs(60)))
+            .await?;
+        self.synchronize_blocks(conn.peer_node_id(), &mut client).await?;
+        Ok(())
+    }
+
+    async fn synchronize_blocks(
+        &mut self,
+        peer: &NodeId,
+        client: &mut rpc::BaseNodeSyncRpcClient,
+    ) -> Result<(), BlockSyncError>
+    {
+        let tip_header = self.blockchain_state.get_tip_header().await?;
+        let local_metadata = self.blockchain_state.get_chain_metadata().await?;
+        if tip_header.height <= local_metadata.height_of_longest_chain() {
+            debug!(
+                target: LOG_TARGET,
+                "Blocks already synchronized to height {}.", tip_header.height
+            );
+            return Ok(());
+        }
+
+        let tip_hash = tip_header.hash();
+        let best_height = local_metadata.height_of_longest_chain();
+        let best_block = self
+            .blockchain_state
+            .get_header(best_height)
+            .await?
+            .ok_or_else(|| BlockSyncError::ExpectedHeaderNotFound(best_height))?;
+        let best_full_block_hash = best_block.hash();
+        debug!(
+            target: LOG_TARGET,
+            "Starting block sync from peer `{}`. Current best block is #{} `{}`. Syncing to #{} ({}).",
+            peer,
+            best_height,
+            best_full_block_hash.to_hex(),
+            tip_header.height,
+            tip_hash.to_hex()
+        );
+        let request = SyncBlocksRequest {
+            start_hash: best_full_block_hash.clone(),
+            // To the tip!
+            end_hash: tip_hash,
+        };
+
+        let mut block_stream = client.sync_blocks(request).await?;
+        let mut prev_hash = best_full_block_hash;
+        let mut current_header = None;
+        while let Some(block) = block_stream.next().await {
+            let block = block?;
+
+            let header = self
+                .blockchain_state
+                .get_header_by_hash(block.hash.clone())
+                .await?
+                .ok_or_else(|| {
+                    BlockSyncError::ReceivedInvalidBlockBody("Peer sent hash for block header we do not have".into())
+                })?;
+
```

Comment:
We  should probably check the height matches at this point as well

Reply:
The headers are trusted at this point (since header sync downloaded and validated them incl the heights) - I do check that the given hash forms a chain according to prev_hash.

Reply:
Yeah I see the prev_hash is  checked, so it's not  possible for the sync peer to send a block out  of order